File: tests/gnutls.pl
#!/usr/bin/perl

my $TMPFILE="out-gnutls.tmp";

if ($ENV{OLD_GNUTLS} eq '1') {
    print "Not checking the GnuTLS configuration\n";
    exit 0
}

print "Checking the GnuTLS configuration\n";

my $dir = 'tests/outputs';

opendir(DIR, $dir) or die $!;

my @gnutlspolicies
    = grep {
        /gnutls/          # has gnutls in name
        && -f "$dir/$_"   # and is a file
    } readdir(DIR);

foreach my $policyfile (@gnutlspolicies) {
	my $policy = $policyfile;
	$policy =~ s/-[^-]+$//;

	print "Checking policy $policy\n";
	next if $policy eq 'GOST-ONLY';

	system("GNUTLS_DEBUG_LEVEL=3 GNUTLS_SYSTEM_PRIORITY_FILE=$dir/$policyfile GNUTLS_SYSTEM_PRIORITY_FAIL_ON_INVALID=1 gnutls-cli -l >$TMPFILE 2>&1");
	if ($? == 0 && $policy eq 'EMPTY') {
		print "Error in gnutls empty policy ($policy)\n";
		system("cat $TMPFILE 1>&2");
		exit 1;
	} elsif ($? != 0 && $policy ne 'EMPTY') {
		print "Error in gnutls policy for $policy\n";
		system("cat $TMPFILE 1>&2");
		exit 1;
	}
	unlink($TMPFILE);
}

exit 0;

File: tests/gnutls.py
#!/usr/bin/python3

import os
import subprocess
import sys
from pathlib import Path

if os.getenv('OLD_GNUTLS') == '1':
    print('Not checking the GnuTLS configuration')
    sys.exit(0)

print('Checking the GnuTLS configuration')

for policy_path in Path('tests', 'outputs').glob('*-gnutls.txt'):
    policy = policy_path.name.removesuffix('-gnutls.txt')
    if policy == 'GOST-ONLY':
        continue
    print(f'Checking policy {policy}')

    p = subprocess.run(['gnutls-cli', '-l'],  # noqa: S607
                       env={**os.environ,
                            'GNUTLS_DEBUG_LEVEL': '3',
                            'GNUTLS_SYSTEM_PRIORITY_FILE': policy_path,
                            'GNUTLS_SYSTEM_PRIORITY_FAIL_ON_INVALID': '1'},
                       check=False,
                       encoding='utf-8',
                       stdout=subprocess.PIPE, stderr=subprocess.STDOUT)

    if p.returncode == 0 and policy == 'EMPTY':
        print(f'Error in gnutls empty policy {policy}', file=sys.stderr)
        print(p.stdout, file=sys.stderr)
        sys.exit(1)
    elif p.returncode != 0 and policy != 'EMPTY':
        print(f'Error in gnutls policy {policy}', file=sys.stderr)
        print(p.stdout, file=sys.stderr)
        sys.exit(1)
